test/basic.test.js

```diff
@@ -23,43 +23,214 @@ test('Connect-Subscribe-Publish-Disconnect 300 clients using WS and MQTT/MQTTS p
   await Promise.all(clients.map(c => c.end()))
 })
 
-test('Subscribed clients receive updates', async function (t) {
-  t.plan(10, 'each client should receive a message')
+test('Subscribed clients receive updates - QoS 1', async function (t) {
+  t.plan(20, 'each client should receive a message')
   t.tearDown(helper.closeBroker)
 
   await helper.startBroker()
 
-  var msg = {
-    topic: 'subscribers/topic',
-    payload: 'Hello world',
-    qos: 1,
-    retain: false
+  async function testQos (qos) {
+    var msg = {
+      topic: 'subscribers/topic',
+      payload: 'Hello world',
+      qos: qos,
+      retain: false
+    }
+
+    var subscribers = []
+    var received = 0
+
+    for (let i = 0; i < 10; i++) {
+      subscribers.push(helper.startClient())
+    }
+
+    subscribers = await Promise.all(subscribers)
+
+    function onMessage (topic, message) {
+      if (topic === msg.topic) {
+        if (received++ < 10 || qos === 2) {
+          t.pass('message received')
+        }
+      }
+    }
+
+    for (let i = 0; i < subscribers.length; i++) {
+      const client = subscribers[i]
+      await client.subscribe(msg.topic)
+      client.once('message', onMessage)
+    }
+
+    var publisher = await helper.startClient()
+
+    await publisher.publish(msg.topic, msg.payload, msg)
+    await helper.delay(100)
+
+    await Promise.all(subscribers.map(c => c.end()))
+    await publisher.end()
   }
 
-  var subscribers = []
+  await testQos(1)
+  await testQos(2)
```

Comment:
This should be 2 tests, not one.

Reply:
The code is the same, in other way I have to move the function outside and create 2 tests